src/pool/srv_target.c

```diff
@@ -1563,3 +1563,201 @@ ds_pool_tgt_query_map_handler(crt_rpc_t *rpc)
+{
+	pool_target_addr_list_free(&arg->tgt_list);
+	D_FREE_PTR(arg);
+}
+
+static int
+obj_discard_cb(daos_handle_t ch, vos_iter_entry_t *ent,
+	       vos_iter_type_t type, vos_iter_param_t *param,
+	       void *data, unsigned *acts)
+{
+	struct child_discard_arg	*arg = data;
+	daos_epoch_range_t		epr;
+	int				rc;
+
+	epr.epr_hi = arg->tgt_discard->epoch;
+	epr.epr_lo = 0;
+	rc = vos_discard(param->ip_hdl, &ent->ie_oid, &epr, NULL, NULL);
+	if (rc != 0)
+		D_ERROR("discard object pool/object "DF_UUID"/"DF_UOID" rc: "DF_RC"\n",
+			DP_UUID(arg->tgt_discard->pool_uuid), DP_UOID(ent->ie_oid),
+			DP_RC(rc));
+	return rc;
+}
+
+/** vos_iter_cb_t */
+static int
+cont_discard_cb(daos_handle_t ih, vos_iter_entry_t *entry,
+		vos_iter_type_t type, vos_iter_param_t *iter_param,
+		void *cb_arg, unsigned int *acts)
+{
+	struct child_discard_arg *arg = cb_arg;
+	struct ds_cont_child	*cont = NULL;
+	vos_iter_param_t	param = { 0 };
+	struct vos_iter_anchors	anchor = { 0 };
+	daos_handle_t		coh;
+	int			rc;
+
+	D_ASSERT(type == VOS_ITER_COUUID);
+	if (uuid_compare(arg->cont_uuid, entry->ie_couuid) == 0) {
+		D_DEBUG(DB_REBUILD, DF_UUID" already discard\n",
+			DP_UUID(arg->cont_uuid));
+		return 0;
+	}
+
+	rc = ds_cont_child_lookup(arg->tgt_discard->pool_uuid, entry->ie_couuid,
+				  &cont);
+	if (rc != DER_SUCCESS) {
+		D_ERROR("Lookup container '"DF_UUIDF"' failed: "DF_RC"\n",
+			DP_UUID(entry->ie_couuid), DP_RC(rc));
+		return rc;
+	}
+
+	rc = vos_cont_open(iter_param->ip_hdl, entry->ie_couuid, &coh);
+	if (rc != 0) {
+		D_ERROR("Open container "DF_UUID" failed: "DF_RC"\n",
+			DP_UUID(entry->ie_couuid), DP_RC(rc));
+		D_GOTO(put, rc);
+	}
+
+	param.ip_hdl = coh;
+	param.ip_epr.epr_lo = 0;
+	param.ip_epr.epr_hi = arg->tgt_discard->epoch;
+	uuid_copy(arg->cont_uuid, entry->ie_couuid);
+
+	rc = vos_iterate(&param, VOS_ITER_OBJ, false, &anchor, obj_discard_cb, NULL,
+			 arg, NULL);
+	vos_cont_close(coh);
+	D_DEBUG(DB_TRACE, DF_UUID"/"DF_UUID" discard cont done: "DF_RC"\n",
+		DP_UUID(arg->tgt_discard->pool_uuid), DP_UUID(entry->ie_couuid),
+		DP_RC(rc));
+
+put:
+	ds_cont_child_put(cont);
+	return rc;
+}
+
+static int
+pool_child_discard(void *data)
+{
+	struct tgt_discard_arg	*arg = data;
+	struct child_discard_arg cont_arg;
+	struct ds_pool_child	*child;
+	vos_iter_param_t	param = { 0 };
+	struct vos_iter_anchors	anchor = { 0 };
+	struct pool_target_addr addr;
+	uint32_t		myrank;
+	int			rc;
+
+	crt_group_rank(NULL, &myrank);
+	addr.pta_rank = myrank;
+	addr.pta_target = dss_get_module_info()->dmi_tgt_id;
+	if (!pool_target_addr_found(&arg->tgt_list, &addr))
+		return 0;
+
+	D_DEBUG(DB_MD, DF_UUID" discard %u/%u\n", DP_UUID(arg->pool_uuid),
+		myrank, addr.pta_target);
+
+	child = ds_pool_child_lookup(arg->pool_uuid);
+	D_ASSERT(child != NULL);
+	param.ip_hdl = child->spc_hdl;
+
+	cont_arg.tgt_discard = arg;
+	rc = vos_iterate(&param, VOS_ITER_COUUID, false, &anchor,
+			 cont_discard_cb, NULL, &cont_arg, NULL);
```

Comment:
[Background question] By the way, how is the reintegration feature handling new containers created and existing containers destroyed while this engine was excluded?

Reply:
container will be created during migrating process, i.e. if the container uuid does not exist, it will open/create the container. see migrate_get_cont_child().  or I miss your question?

Reply:
That answers the "containers created" part. Thanks. What about "containers destroyed"?

Reply:
If the container is being destroyed, then it will return -DER_NONEXIST, see obj_ioc_init(), and rebuild migration will ignore ENOEXIST error during enumeration and fetch. 

Probably there is better way to handle it, but any rebuild failure will cause rebuild to retry, so any of these temporary error will be "fixed" in the next retry,  so I did not try very hard here.

Reply:
In addition to ignoring -DER_NONEXISTs, how does the reintegration protocol release the space consumed by "destroyed containers" eventually?

Reply:
Reintegration/rebuild does not destroy the container, and it might hold these orphan container, which will be released(destroyed) eventually. 

Hmm, not sure I follow your question here, could you please elaborate  a bit? thanks.

Reply:
For example, I have a DAOS pool in which every target is 1 GB. I create and write to a container, so that it consumes about 512 MB on every target. An engine crashes and is excluded from the pool. I destroy this container so that its 512 MB data on every remaining target is released. I restart and reintegrate the excluded engine. How will the container's 512 MB data be destroyed in each of the target on this reintegrated engine?

Reply:
>How will the container's 512 MB data be destroyed in each of the target on this reintegrated engine?

oh, before reintegration, it will discard all the data on the reintegrating target (ds_pool_tgt_discard_handler),  which will destroy this orphan container anyway.

Reply:
I see. We do not support reintegrating an engine preserving existing data, correct?

Reply:
right, not yet.

Reply:
OK, thanks.